### DISCUSSION TITLE
Typing list comprehensions (Py 3.9 and Py 3.10)

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1048] [Q&A] [answered]
What is the correct way to type list comprehensions? For example:

my_numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]

my_squares = [x + sqr(x) for x in my_numbers]

**Comment by srittau** [comment]
The result of a list comprehension is simply a list. Often the type can be determined automatically by a type checker, but an explicit annotation is also possible:

```python
my_numbers: list[int] = [1, 2, 3, 4, 5, 6, 7, 8, 9]
my_squares: list[float] = [x + sqr(x) for x in my_numbers]
```

This requires `from __future__ import annotations` in Python 3.9.

**Reply by sobolevn** [reply]
I would add that annotations here are completely optional:

```python
my_numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
reveal_type(my_numbers)  # N: Revealed type is "builtins.list[builtins.int*]"

my_squares = [x + x for x in my_numbers]
reveal_type(my_squares)  # N: Revealed type is "builtins.list[builtins.int*]"
```

Most type-checkers will just infer the correct type.

**Comment by dineshbvadhia** [comment, marked as answer]
I was asking about the x in the list comprehension?

**Reply by srittau** [reply]
The `x` can't be annotated directly. If `my_numbers` is annotated correctly or the type is inferred correctly (you can check that with `reveal_type`), `x` should also be inferred correctly. For example, if `my_numbers` is `list[int]`, type checkers infer correctly that `x` is `int`.
